### PR TITLE
Secondary emission on domain sides for JSON model

### PR DESCRIPTION
**Pull request by FannySkirbekk**
Added parseDomainReactions(), side-dependency is yet to be implemented.

### CODE REVIEW

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.cpp [hunk 1]
```diff
@@ -1,12 +1,12 @@
 /* chombo-discharge
- * Copyright © 2021 SINTEF Energy Research.
+ * Copyright © 2022 SINTEF Energy Research.
```
Comment (rmrsk) [inline Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.cpp:2]: Add Fanny or NTNU copyright.

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H [hunk 2]
```diff
@@ -1,12 +1,12 @@
 /* chombo-discharge
- * Copyright © 2021 SINTEF Energy Research.
+ * Copyright © 2022 SINTEF Energy Research.
```
[the remark this answers is not in the document]
Comment (rmrsk) [reply Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:2]: Below: E.g. '* Copyright @ 2022 NTNU'

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
+ * Copyright © 2022 NTNU.
+ * Copyrigth © 2022 Fanny Skirbekk.
```
Comment (rmrsk) [inline Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:4]: Typo

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H [hunk 4]
```diff
@@ -647,7 +649,46 @@ namespace Physics {
...
       */
-      std::vector<CdrPlasmaSurfaceReactionJSON> m_dielectricReactions;      
+      std::vector<CdrPlasmaSurfaceReactionJSON> m_dielectricReactions;
+
+       // ===========================================
+      // DOMAIN SURFACE REACTIONS BEGIN HERE
+      // ===========================================
+
+      /*!
+	@brief Lookup method for the domain reaction ratesThe pair is made up of an int representing direction (0=x, 1=y, 2=z) and a Side::LoHiSide
```
Comment (rmrsk) [inline Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:659]: ratesThe pair => rates. The pair

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H [hunk 5]
```diff
@@ -647,7 +649,46 @@ namespace Physics {
...
+	representing side (Side::Lo, Side::Hi)
+      */
+      std::map<std::pair<int, Side::LoHiSide>, std::map<int, LookupMethod>> m_domainReactionLookup;
```
Comment (rmrsk) [inline Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:662]: To be kind to older compilers, let '>>' be separate by a space. I.e. '> >'. Older compilers will complain.

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H [hunk 6]
```diff
@@ -647,7 +649,46 @@ namespace Physics {
...
+      /*! 
+	@brief List of domain reactions. The pair is made up of an int representing direction (0=x, 1=y, 2=z) and a Side::LoHiSide
+	representing side (Side::Lo, Side::Hi)
+      */
+      std::map<std::pair<int, Side::LoHiSide>, std::vector<CdrPlasmaSurfaceReactionJSON>> m_domainReactions;
+
+      /*! 
+	@brief map to translate dir from char to int
+      */
+      const std::map<char, int> m_dirCharToInt{{'x', 0}, {'y', 1}, {'z', 2}};
```
Comment (rmrsk) [inline Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:686]: Does this have to be a member? If yes, make it constexpr?

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H [hunk 7]
```diff
@@ -647,7 +649,46 @@ namespace Physics {
...
+
+      /*! 
+	@brief map to translate side from std::string to Side::LoHiSide
+      */
+      const std::map<std::string, Side::LoHiSide> m_sideStringToSide{{"lo", Side::Lo}, {"hi", Side::Hi}};
```
Comment (rmrsk) [inline Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:691]: Does this have to be a member? If yes, make it constexpr?

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H [hunk 8]
```diff
@@ -875,7 +916,30 @@ namespace Physics {
...
+
+      /*!
+	@brief Parse secondary emission on domain
+      */
+      virtual void parseDomainReactions();
+
+      /*!
+	@brief Parse reaction rate for domain reactions. 
+	@param[in] a_reactionIndex  Reaction index. 
+	@param[in] a_reactionJSON   Input reaction. Must be one of the entries in the 'electrode reactions' array.
```
Comment (rmrsk) [inline Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:929]: 'electrode reactions' => 'domain reactions'

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H [hunk 9]
```diff
@@ -875,7 +916,30 @@ namespace Physics {
...
+	@param[in] a_sides          List of the domain sides that this occurs at
+      */
+      virtual void parseDomainReactionRate(const int a_reactionIndex, const json& a_reactionJSON, const std::vector<std::string>& a_sides);      
+
+      /*!
+	@brief Parse domain reaction scaling for a specific reaction.
+	@details This currently exists without much functionality -- it just returns one for all reactions. As we
```
Comment (rmrsk) [inline Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:936]: Not true..?

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.cpp [hunk 10]
```diff
@@ -1,12 +1,14 @@
...
+ * Copyright © 2022 NTNU.
+ * Copyrigth © 2022 Fanny Skirbekk.
```
Comment (rmrsk) [inline Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.cpp:4]: Typo

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.cpp [hunk 11]
```diff
@@ -2455,6 +2458,182 @@ void CdrPlasmaJSON::parseDielectricReactionScaling(const int a_reactionIndex, co
...
+
+	// Make the string-int encoding so we can encode the reaction properly. Then add the reaction to the pile.
+	std::list<int> plasmaReactants;
+	std::list<int> neutralReactants;
+	std::list<int> photonReactants;
+	std::list<int> plasmaProducts;
+	std::list<int> neutralProducts;
+	std::list<int> photonProducts;
+	
+
```
Comment (rmrsk) [inline Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.cpp:2528]: Delete line.

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.cpp [hunk 12]
```diff
@@ -2455,6 +2458,182 @@ void CdrPlasmaJSON::parseDielectricReactionScaling(const int a_reactionIndex, co
...
+  }
+  
+  // Check if JSON file contains domain reactions
+  if(m_json.contains("domain reactions")){
+    const json& domainReactions = m_json["domain reactions"];
+
+    // Base error when parsing the reactions
+    const std::string baseError = "CdrPlasmaJSON::parseDomainReactions ";
+
+    //const std::map<char, int> dirMap{{'x', 0}, {'y', 1}, {'z', 2}};
```
Comment (rmrsk) [inline Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.cpp:2474]: Delete if not needed.

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.cpp [hunk 13]
```diff
@@ -2455,6 +2458,182 @@ void CdrPlasmaJSON::parseDielectricReactionScaling(const int a_reactionIndex, co
...
+  // We MUST have a field lookup because it determines how we compute the efficiencies for surface reactions
+  if(!(a_reactionJSON.contains("lookup"))) this->throwParserError(baseError + "but field 'lookup' was not specified");
+
+  // Get the lookup method
+  const std::string lookup = this->trim(a_reactionJSON["lookup"].get<std::string>());
+
+  // Now go through the various rate-computation methods and fill the relevant containers
+  if(lookup == "constant"){
+    // If using a constant emission rate, we must get the field 'value'
+
```
Comment (rmrsk) [inline Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.cpp:2587]: Delete line.

**Review by rmrsk — APPROVED** [review]

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H (hunk 6 above)
Comment (FannySkirbekk) [reply Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:686]: initially I made it as a local variable in the CdrPlasmaJSON::parseDomainReactions()-function, but I changed it to a member when I needed it in both parseDomainReactionRate() and parseDomainReactionScaling() as well. The alternative would be to have it as a variable in all three functions, which option do you prefer?

And yes, constexpr is a good idea if it stays as a member

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H (hunk 9 above)
Comment (FannySkirbekk) [reply Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:936]: I agree. I'll remove it for electric and dielectric as well

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.cpp [hunk 14]
```diff
@@ -2236,7 +2236,7 @@ void CdrPlasmaJSON::parseElectrodeReactions() {
 	this->sanctifySurfaceReaction(curReactants, curProducts, reaction);
 
 	// This is the reaction index for the current index. The reaction we are currently
-	// dealing with is put in m_plasmaReactions[reactionIdex]. 
+	// dealing with is put in m_electrodeReactions[reactionIndex].
```
Comment (FannySkirbekk) [inline Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.cpp:2239]: Is my version correct? If yes: I'll fix it for dielectric as well, if no: I'll change it back

File: Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H (hunk 6 above)
Comment (rmrsk) [reply Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:686]: Used more than once -> member.

Comment (FannySkirbekk) [reply Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:686]: Do you want me to place it somewhere else in the .h-file instead, or is it ok to keep it as is?

Comment (FannySkirbekk) [reply Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:686]: It seems like std::map cannot be constexpr, only const

Comment (rmrsk) [reply Physics/CdrPlasma/PlasmaModels/CdrPlasmaJSON/CD_CdrPlasmaJSON.H:686]: Ok, keep as is.

**Review by rmrsk — DISMISSED** [review]

**Review by rmrsk — CHANGES_REQUESTED** [review]
Looks good. 

Reset positive2d.inputs and I will merge.

**Comment by FannySkirbekk** [comment]
Tried resetting positive2d.inputs, does it look ok?

**Review by rmrsk — APPROVED** [review]

**Pull request merged**